packages/flutter/lib/src/services/asset_vendor.dart

```diff
@@ -0,0 +1,55 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+
+import 'package:mojo/core.dart' as core;
+
+import 'asset_bundle.dart';
+import 'asset_resolver.dart';
+
+/// Builder for a particular AssetResolver.
+///
+/// Since AssetVendor uses builder objects as cache keys, subclasses of
+/// AssetResolverBuilder should override operator== and hashCode so that
+/// equivalent configurations appear to be equal. Builder objects should also be
+/// cheap to construct.
+abstract class AssetResolverBuilder<T extends AssetResolver> {
+  const AssetResolverBuilder();
+  T buildResolver();
+}
+
+/// Asset bundle capable of vending assets via the resolution logic of an
+/// [AssetResolver].
+///
+/// Each AssetVendor is specific to a particular underlying asset bundle and
+/// asset resolver configuration. To retrieve an appropriate AssetVendor, call
+/// the static `vendorFor` method.
+class AssetVendor extends CachingAssetBundle {
+
+  static Map<AssetBundle, Map<AssetResolverBuilder, AssetVendor>> _cache =
+    <AssetBundle, Map<AssetResolverBuilder, AssetVendor>>{};
+
+  /// Retrieve an asset vendor for the given asset bundle and asset resolver
+  /// configuration. If no such vendor exists yet, one will be created via
+  /// the configuration's `createResolver` method.
+  static AssetVendor vendorFor(AssetBundle bundle, AssetResolverBuilder builder) {
+    return _cache.putIfAbsent(bundle, () {
+      return <AssetResolverBuilder, AssetVendor>{};
+    }).putIfAbsent(builder, () {
+      return new AssetVendor._(bundle: bundle, resolver: builder.buildResolver());
+    });
+  }
+
+  AssetVendor._({ this.bundle, this.resolver });
+  final AssetBundle bundle;
+  final AssetResolver resolver;
+
+  Map<String, String> _keyCache = <String, String>{};
+
+  Future<core.MojoDataPipeConsumer> load(String key) async {
+    if (!_keyCache.containsKey(key))
```

Comment:
You can structure this code so that it only does a single map lookup in the cache hit case by using `operator[]`.  If the result is null, then you need to populate the cache.  Otherwise, you can just use the value you get back.